Added deconstruction function and other fixes

brodo97 opened the pull request:
Added **deconstruct** function. Now you can deconstruct buildings level by level.
I found that useful on my mines upgrade schedule. Once the planet produce a good amount of deuterium then it can convert the energy production from solar to fusion and therefore consume less spaces

Fixed **spyreports**' ugliness. More elegant, compact and smart
Fixed stuff like missing spaces and quotation marks

#117 
Fixed **friendly_fleet** and **hostile_fleet** crash on id casting
Can this fix the problem and also do not destroy all running bots?

brodo97 committed:
Some modifications to __init__.py

Added deconstruct function. Now you can deconstruct buildings level by level
Fixed spyreports' ugliness. More elegant, compact and smart
Fixed stuff like missing spaces and quotation marks

Fixed friendly_fleet and hostile_fleet crash on id casting

brodo97 committed:
Update test.py

Added deconstruct test

brodo97 committed:
Update README.md

Added deconstruct explanation

PiecePaperCode commented on ogame/__init__.py:
```diff
@@ -1154,64 +1154,49 @@ def spyreports(self, firstpage=1, lastpage=30):
-                spied_research.update({tech_name: tech_amount})
+                resources_data[resource_name[0]] = int(resource['title'].replace('.', ''))
+
+            def get_tech_qt(tech_type):
```
what does qt stand for?
needs bedder name

brodo97 replied:
It stands for quantity. I'm gonna rename that

PiecePaperCode commented on ogame/__init__.py:
```diff
@@ -1154,64 +1154,49 @@ def spyreports(self, firstpage=1, lastpage=30):
+                'buildings': [const.buildings.building_name, None],
+                'research': [const.research.research_name, 'research']
+            }
+            for tech_type in spied_data.keys():
```
O(n2) complexity.
can this be reduced to O(n)?

brodo97 replied:
Sorry, I never understood that O(x) thing...

I looked on Wikipedia, now I sort of understand what O(x) mean but I can't find where O(n<sup>2</sup>) is in my code. Can you please pointing it out so I can better understand?

PiecePaperCode replied:
i am pointing out that you have a quite complex algo.
for
    for
responds to a O(n2)
and when possible its best to avoid doing it like this because its quite scales exponential. depends on the runtime it will run against you.

if i have 20 spy messages it will run 20^2=400 times

i try to aboid loops inside loops

brodo97 replied:
Thanks for the explanation!
You're right..

Now the structure is:
<pre>
for (every message)
    for (tech type)
        for (every tech)
</pre>

But I'm not quite sure how to fix it.. For every message I still have to cycle all tech types and all techs.

I'll try to optimize it, maybe I can read some guides!

PiecePaperCode replied:
ill try something out later maby i could give a try optemising it

PiecePaperCode commented on ogame/__init__.py:
```diff
@@ -1310,6 +1295,30 @@ def build(self, what, id):
                     'menge': amount}
         )
 
+    def deconstruct(self, what, id):
```
really clean implementation. easy to understand and how it is structured good job

PiecePaperCode commented on ogame/test.py:
```diff
@@ -174,6 +174,20 @@ def test_build(self):
         ).rocket_launcher
         self.assertTrue(before < after.amount or after.in_construction)
 
+    def test_deconstruct(self):
```
is it possible to use somthing much cheaper to deconstrukt or at least cancel it. would produce unwanted side effects. losing my 10 mio metal mine when i test my code

brodo97 replied:
Yeah... I was thinking about that too. But I don't know what to use instead. Any advice?
I was doing my tests with _buildings.missile_silo_ on the only planet where I have it at level 1

Also, I noticed now that I'm calling a tuple on line 183: `what=buildings.metal_mine(),`. Going to fix that..

brodo97 replied:
Locally I added the cancel_***something***() function. I'm now quite sure how to call it... It can cancel construction and deconstruction. It can also cancel research. And I'm pretty sure that you can cancel shipyard queue when you have the commander, but I can't test it now so I can't implement it.

So we'll be able to test the ***deconstruct()*** function without trashing millions of resources

PiecePaperCode reviewed (changes requested):
good job i loke your effort alot. i have some small changes i would love to see. il aprove them right away

brodo97 committed:
New cancel function and some fixes

Added cancel_building function to cancel construction/deconstruction of buildings and cancel_research to cancel the ongoing research. (Didn't test with commander)
Fixed celestial function temperature extraction. Instead of trying a list of possible regex we can extract the two numbers in the textContent[3] var using another regex.

brodo97 committed:
Less stuff, less problems

brodo97 committed:
Added cancel function testing

Should I also add the research test?

brodo97 committed:
New cancel function explanation

PiecePaperCode commented on ogame/__init__.py:
```diff
@@ -1310,6 +1289,62 @@ def build(self, what, id):
+                    'type': type}
+        )
+
+    def cancel_building(self, id):
```
they are only used by cancel you could nest them instead of making them visible by the user. but its fine also as is

PiecePaperCode commented on ogame/test.py:
```diff
@@ -174,6 +174,32 @@ def test_build(self):
+        ).metal_mine
+        self.assertTrue(before.level > after.level or after.in_construction)
+
+    def check_cancel(self):
```
the test framework dosent run in a row usualy. if one start before the other the cancel test will fail. ether you combine the two or you implement them to eachother. not really a problem as it can be changed easy

brodo97 replied:
Maybe the best option is to test both simultaneously.

PiecePaperCode reviewed (approved):
small issues left that dont need to be adressed.

good job love the work you put in
